File: 1112-find-words-that-can-be-formed-by-characters/solution.cpp
#include <iterator>
#include <map>

class Solution {
public:
    int countCharacters(vector<string>& words, string chars) {
        
        int result = 0;
    
        for(int j = 0; j < words.size(); j++)
        {
            // cout<< words[j] << endl;
            
            // create a map
            map<char, int> char_map;
            for(int i = 0; i < chars.length(); i++)
            {
                char_map[chars[i]] = 0;
            }
            for(int i = 0; i < chars.length(); i++)
            {
                char_map[chars[i]] += 1;
            }
            
            bool is_in = true;
            
            for(int k = 0; k < words[j].length(); k++)
            {
                //find the element in the map
                if(char_map.find(words[j][k]) != char_map.end())
                {
                    if(char_map[words[j][k]] > 0)
                        char_map[words[j][k]] -= 1;    
                    else
                    {
                        is_in = false;
                        break;
                    }
                }
                
                else
                {
                    is_in = false;
                    break;
                }
                
            }
            // cout<< is_in<<endl;
            if(is_in)
            {
                result += words[j].length();
            }
        }
        
        return result;

    }
};

File: 1112-find-words-that-can-be-formed-by-characters/solution.py
class Solution:
    def countCharacters(self, words: List[str], chars: str) -> int:
        result = 0
        for word in words:
            is_in = True
            c_n = {}
            for char in chars:
                if char in c_n:
                    c_n[char] += 1
                else:
                    c_n[char] = 1
            
            for c in word:
                if c in c_n:
                    if c_n[c] > 0:
                        c_n[c] -= 1
                    else:
                        is_in = False
                        break
                else:
                    is_in = False
                    break
            if is_in:
                result += len(word)
            else:
                continue
        return result
